pageserver/src/tenant/timeline.rs

```diff
@@ -1015,49 +1015,69 @@ impl Timeline {
-                                if let Err(err) = rename_to_backup(&local_layer_path) {
-                                    error!("could not rename file {local_layer_path:?}: {err:?}");
-                                } else {
-                                    self.metrics.current_physical_size_gauge.sub(local_size);
-                                    already_downloaded = false;
-                                }
-                            }
-                        }
-                        Err(err) => {
-                            error!("could not get size of local file {local_layer_path:?}: {err:?}")
+                    let metadata = local_layer_path.metadata().with_context(|| {
+                        format!(
+                            "get file size of local layer {}",
+                            local_layer_path.display()
+                        )
+                    })?;
+                    let local_size = metadata.len();
+                    if local_size != remote_size {
+                        warn!("removing local file {local_layer_path:?} because it has unexpected length {local_size}; length in remote index is {remote_size}");
+                        if let Err(err) = rename_to_backup(&local_layer_path) {
+                            anyhow::bail!("could not rename file {local_layer_path:?}: {err:?}");
+                        } else {
+                            self.metrics.current_physical_size_gauge.sub(local_size);
+                            assert!(!local_layer_path.exists());
+                            self.layers.write().unwrap().remove_historic(local_layer);
+                            // fall-through to adding the remote layer
                         }
+                    } else {
+                        info!(
+                            "layer is present locally and file size matches remote, using it: {}",
+                            local_layer_path.display()
+                        );
+                        continue;
                     }
-                }
-
-                if already_downloaded {
+                } else {
+                    info!(
+                        "layer is present locally and remote does not have file size, using it: {}",
+                        local_layer_path.display()
+                    );
                     continue;
                 }
             } else {
-                info!("remote layer {remote_layer_name:?} does not exist locally");
+                let local_layer_path = timeline_dir.join(remote_layer_name.file_name());
+                assert!(
+                    !local_layer_path.exists(),
+                    "caller did not supply exhaustive list of local_layers: {local_layer_path:?}"
+                );
+                // fall-through to adding the remote layer
             }
 
+            info!(
```

Comment:
Assuming you will do the demotions from `info` => `debug` above, and we see only these log lines in the log, not the download progress, I think the outcome will be confusing.

I'd propose adding the next action to this message ("downloading"), and maybe an overall completion message with how many were downloaded before return?